### DISCUSSION TITLE
Sloc Cloc and Code (scc) analysis of repo claims $4M cost to develop Athena++

### DISCUSSION THREAD

**Discussion by felker** [PrincetonUniversity/athena#633] [General]
I often use [`sloc`](https://github.com/flosse/sloc) to track the growth of repos and keep an eye on code bloat. As @jmstone says, once you get around ~100K lines of C/C++, the project can start to become unwieldy (at least for computational science codes supported by folks from only a handful of academic and research institutions). `sloc` was really useful for some rewrites of the `bvals/` content, etc. 

Recently learned about a much more powerful extension of a code counter, https://github.com/boyter/scc, and thought I would share it here for fun. Here is the basic output:
```
❯ scc .
───────────────────────────────────────────────────────────────────────────────
Language                 Files     Lines   Blanks  Comments     Code Complexity
───────────────────────────────────────────────────────────────────────────────
C++                        288    119623    10894     19185    89544      23220
Python                     136     21142     2182      3819    15141       2979
C++ Header                  70      9121     1365      2045     5711         51
Shell                       11       999      108       350      541         96
Markdown                     8      1293      255         0     1038          0
C Header                     6       353       62       110      181          0
XML                          5       327        0         0      327          0
Autoconf                     3      2870      634      1756      480          4
License                      2       379       66         0      313          0
YAML                         2       124       19        43       62          0
C                            1       864      185       134      545        205
Plain Text                   1       368       78         0      290          0
───────────────────────────────────────────────────────────────────────────────
Total                      533    157463    15848     27442   114173      26555
───────────────────────────────────────────────────────────────────────────────
Estimated Cost to Develop (organic) $3,908,741
Estimated Schedule Effort (organic) 23.09 months
Estimated People Required (organic) 15.04
───────────────────────────────────────────────────────────────────────────────
Processed 6058556 bytes, 6.059 megabytes (SI)
───────────────────────────────────────────────────────────────────────────────
```

Uses the [COCOMO](https://en.wikipedia.org/wiki/COCOMO) software cost estimation model and computes cyclomatic complexity for all the files. Discussion from the project author about the latter here: https://github.com/boyter/scc/discussions/235
It obviously doesn't take into account that some parts of the repo benefitted from being indirectly ported from https://github.com/PrincetonUniversity/Athena-Cversion, and ZEUS etc. before that. And the 4x constants from the "organic" type project for the Basic COCOMO model [here](https://github.com/boyter/scc/blob/master/processor/cocomo.go) could probably be tuned--- plus, it assumes an average salary of $56,286 and an overhead of 2.4 in 
https://github.com/boyter/scc/blob/c55ec6a0fc06f8a69a46968001801e2b3c89901f/processor/cocomo.go#L31
... which strikes me as a bit high for some of the grad students who worked on the code, speaking from experience

Tons of fun options to play around with and get insights on the codebase. For example, ranking the files by complexity:
```
❯ scc --by-file -s complexity
───────────────────────────────────────────────────────────────────────────────
Language                 Files     Lines   Blanks  Comments     Code Complexity
───────────────────────────────────────────────────────────────────────────────
C++                        288    119623    10894     19185    89544      23220
───────────────────────────────────────────────────────────────────────────────
~als/orbital/bvals_orbital.cpp      3164       67       224     2873       1009
~task_list/time_integrator.cpp      3080      260       292     2528        965
src/bvals/fc/bvals_fc.cpp           1593       98       120     1375        924
~als/fc/flux_correction_fc.cpp      1749       49       203     1497        921
~rc/chemistry/network/kida.cpp      2288      111       259     1918        725
src/mesh/mesh.cpp                   2237      217       272     1748        571
~ultigrid/multigrid_driver.cpp      2313      199       167     1947        552
src/bvals/cc/mg/bvals_mg.cpp        1271      107       108     1056        459
src/bvals/cc/bvals_cc.cpp           1122       78        95      949        389
~c/task_list/sts_task_list.cpp       999       76        80      843        377
src/outputs/outputs.cpp             1677      108       175     1394        353
src/mesh/weighted_ave.cpp            823       15        84      724        330
src/mesh/amr_loadbalance.cpp        1225      112       155      958        308
~rc/reconstruct/ppm_simple.cpp      1556      154       329     1073        301
src/bvals/bvals_base.cpp             740       40        88      612        288
~vection/orbital_advection.cpp       874       48        86      740        275
src/pgen/read_vtk.cpp                902       47        95      760        239
~integrators/rad_transport.cpp       841       78        98      665        236
```
So @tomo-ono, for your prize of writing the most complex file, you can expect a check for your cut of the $4M soon 😄

**Comment by jmstone** [comment]
Thanks, Kyle, I never knew there were such tools and I find these results eye-opening.  That $4M cost estimate is an interesting number that could be included in future proposals.  The federal funding agencies should know what value they are getting; the total funds awarded for code development over the years does not come close to that level.  Instead it was all on the backs of lowly paid grad students like yourself!  It would be interesting to run scc on AthenaK...

**Comment by boyter** [comment]
I know this is a bit late but if there is anything you need added to scc to assist with what you are doing, or just general questions please do reach out. I am always happy to assist.
